dnf-behave-tests/features/repoquery/globs.feature

```diff
@@ -0,0 +1,180 @@
+      toppler-1:1.0-1.x86_64
+      """
+
+Scenario: repoquery top?d
+ When I execute dnf with args "repoquery top?d"
+ Then the exit code is 0
+  And stdout is
+      """
+      toped-1:1.0-1.src
+      toped-1:1.0-1.x86_64
+      """
+
+Scenario: repoquery top?{d,it}
+ When I execute dnf with args "repoquery top?{{d,it}}"
+ Then the exit code is 0
+  And stdout is
+      """
+      toped-1:1.0-1.src
+      toped-1:1.0-1.x86_64
+      topgit-1:1.0-1.src
+      topgit-1:1.0-1.x86_64
+      topgit-1:1.17.6-1.src
+      topgit-1:1.17.6-1.x86_64
+      """
+
+Scenario: repoquery top[a-f]d
+ When I execute dnf with args "repoquery top[a-f]d"
+ Then the exit code is 0
+  And stdout is
+      """
+      toped-1:1.0-1.src
+      toped-1:1.0-1.x86_64
+      """
+
+Scenario: repoquery top[a-fg]*
+ When I execute dnf with args "repoquery top[a-fg]*"
+ Then the exit code is 0
+  And stdout is
+      """
+      toped-1:1.0-1.src
+      toped-1:1.0-1.x86_64
+      topgit-1:1.0-1.src
+      topgit-1:1.0-1.x86_64
+      topgit-1:1.17.6-1.src
+      topgit-1:1.17.6-1.x86_64
+      """
+
+Scenario: repoquery top[^a-g]*
+ When I execute dnf with args "repoquery top[^a-g]*"
+ Then the exit code is 0
+  And stdout is
+      """
+      toppler-1:1.0-1.src
+      toppler-1:1.0-1.x86_64
+      """
+
+Scenario: repoquery top{ed,pler}
+ When I execute dnf with args "repoquery top{{ed,pler}}"
+ Then the exit code is 0
+  And stdout is
+      """
+      toped-1:1.0-1.src
+      toped-1:1.0-1.x86_64
+      toppler-1:1.0-1.src
+      toppler-1:1.0-1.x86_64
+      """
+
+Scenario: repoquery top[!n-z]{d,aaa,it}
+ When I execute dnf with args "repoquery top[!n-z]{{d,aaa,it}}"
+ Then the exit code is 0
+  And stdout is
+      """
+      toped-1:1.0-1.src
+      toped-1:1.0-1.x86_64
+      topgit-1:1.0-1.src
+      topgit-1:1.0-1.x86_64
+      topgit-1:1.17.6-1.src
+      topgit-1:1.17.6-1.x86_64
+      """
+
+Scenario: repoquery *top[-a-f]*
+ When I execute dnf with args "repoquery *top[-a-f]*"
+ Then the exit code is 0
+  And stdout is
+      """
+      desktop-utils-1:1.0-1.src
+      desktop-utils-1:1.0-1.x86_64
+      desktop-utils-1:1.23.9-1.src
+      desktop-utils-1:1.23.9-1.x86_64
+      toped-1:1.0-1.src
+      toped-1:1.0-1.x86_64
+      """
+
+
+# <name-version> globs
+Scenario: repoquery *top[-a-f]*-1.0
+ When I execute dnf with args "repoquery *top[-a-f]*-1.0"
+ Then the exit code is 0
+  And stdout is
+      """
+      desktop-utils-1:1.0-1.src
+      desktop-utils-1:1.0-1.x86_64
+      toped-1:1.0-1.src
+      toped-1:1.0-1.x86_64
+      """
+
+Scenario: repoquery *top[-a-f]*-1.[1-4]*
+ When I execute dnf with args "repoquery *top[-a-f]*-1.[1-4]*"
+ Then the exit code is 0
+  And stdout is
+      """
+      desktop-utils-1:1.23.9-1.src
+      desktop-utils-1:1.23.9-1.x86_64
+      """
+
+# TODO this seems to incorrectly match many more packages
+@xfail
+Scenario: repoquery *top[-a-f]*-1.[!1-4]*
+ When I execute dnf with args "repoquery *top[-a-f]*-1.[!1-4]*"
```

Comment:
Subject considers '-' in the [!1-4] range as NEVRA separator, which probably causes the problem.
```
>>> import dnf
>>> subj = dnf.subject.Subject("*top[-a-f]*-1.[!1-4]*")
>>> for i in subj.get_nevra_possibilities():
...     print(i.name, i.epoch, i.version, i.release, i.arch)
... 
*top[-a-f]*-1.[!1-4]* None None None None
*top[-a-f]* None 1.[!1 4]* None
*top[-a-f]*-1.[!1 None 4]* None None
# ^^^ last 2 lines are most likely wrong
```

Reply:
Yes, funny thing is the correct variant is not actually there at all. I still haven't looked up exactly where the matching goes wrong here, but in general, I think we have a problem.

The forms are matched as regular expressions and it is not (AFAIK) possible to handle dashes nested inside `[]` as a special case in regexes. I think we have two options in general:

1. use something more powerful for the regexes, like [boost xpressive](https://github.com/lukash/librepo/commit/8d4dd9f8618fa64b37d17c0e65e9180294fa9d71), which seems to have some facility for nested brackets and more. But I'm not sure how viable it is to use boost. And it doesn't look to be best for the job anyway, so bringing in boost just for that isn't great either.

2. Write a parser that will parse the NEVRAs. It could have potential benefits of better matching than the somewhat crude regexes, but wouldn't probably mesh well with the current "forms" interface. Also, it's somewhat nontrivial...

Reply:
Actually, I think I have an idea how to work around it. I'll give it a go later.

Reply:
So actually this test in particular was not affected by the faulty regex parsing. The regex forms didn't match anything and there's a last resort branch that tries to match the argument against the full NEVRA of the packages as-is. The pattern was actually matched correctly, the last bit was matching against the architecture which I didn't think of. I've fixed the test and pushed it as a separate commit temporarily (so that it is clear what I changed), I'll squash it later.